compiler/rustc_ast_passes/src/feature_gate.rs

```diff
@@ -274,10 +274,12 @@ impl<'a> PostExpansionVisitor<'a> {
                 );
             }
             abi => {
-                self.sess.parse_sess.span_diagnostic.delay_span_bug(
-                    span,
-                    &format!("unrecognized ABI not caught in lowering: {}", abi),
-                );
+                if self.sess.opts.pretty.map_or(true, |ppm| ppm.needs_hir()) {
+                    self.sess.parse_sess.span_diagnostic.delay_span_bug(
```

Comment:
Question: Do we _want_ to error when we have an invalid ABI and we don't go thru HIR lowering, like with `-Zunpretty=expanded`? 

Since we're not going thru HIR lowering, we don't emit the error at the normal place we expect it, but if we want, we could emit instead of delay an error here for the earlier pretty modes.

Right now this just allows the invalid ABI string, but I have no strong opinions either way.

Reply:
Per [the reference](https://doc.rust-lang.org/nightly/reference/items/functions.html) there should be no ast-level constraint for abi. I think it is fine to leave it as is in ast.